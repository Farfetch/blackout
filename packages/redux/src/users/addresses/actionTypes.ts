/**
 * Action type dispatched when the create address request fails.
 */
export const CREATE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the create address request starts.
 */
export const CREATE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the create address request succeeds.
 */
export const CREATE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the fetch address request fails.
 */
export const FETCH_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the fetch address request starts.
 */
export const FETCH_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the fetch address request succeeds.
 */
export const FETCH_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the fetch addresses request fails.
 */
export const FETCH_USER_ADDRESSES_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_FAILURE';
/**
 * Action type dispatched when the fetch addresses request starts.
 */
export const FETCH_USER_ADDRESSES_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_REQUEST';
/**
 * Action type dispatched when the fetch addresses request succeeds.
 */
export const FETCH_USER_ADDRESSES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_SUCCESS';

/**
 * Action type dispatched when the fetch default contact address request fails.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when the fetch default contact address request starts.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when the fetch default contact address request succeeds.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the delete address request fails.
 */
export const REMOVE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the delete address request starts.
 */
export const REMOVE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the delete address request succeeds.
 */
export const REMOVE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when delete default contact address request fails.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when delete default contact address request starts.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when delete default contact address request succeeds.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default billing address request fails.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default billing address request starts.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default billing address request succeeds.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default shipping address request fails.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default shipping address request starts.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default shipping address request succeeds.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default contact address request fails.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default contact address request starts.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default contact address request succeeds.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the update address request fails.
 */
export const UPDATE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the update address request starts.
 */
export const UPDATE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the update address request succeeds.
 */
export const UPDATE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when resetting the addresses.
 */
export const RESET_USER_ADDRESSES =
  '@farfetch/blackout-redux/RESET_USER_ADDRESSES';
