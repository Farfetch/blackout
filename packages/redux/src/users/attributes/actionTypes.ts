/**
 * Action type dispatched when the fetch user attributes request fails.
 */
export const FETCH_USER_ATTRIBUTES_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_FAILURE';
/**
 * Action type dispatched when the fetch user attributes request starts.
 */
export const FETCH_USER_ATTRIBUTES_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_REQUEST';
/**
 * Action type dispatched when the fetch user attributes request succeeds.
 */
export const FETCH_USER_ATTRIBUTES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_SUCCESS';

/**
 * Action type dispatched when the create user attributes request fails.
 */
export const CREATE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the create user attributes request starts.
 */
export const CREATE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the create user attributes request succeeds.
 */
export const CREATE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the fetch user attribute request fails.
 */
export const FETCH_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the fetch user attribute request starts.
 */
export const FETCH_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the fetch user attribute request succeeds.
 */
export const FETCH_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the set user attribute request fails.
 */
export const SET_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the set user attribute request starts.
 */
export const SET_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the set user attribute request succeeds.
 */
export const SET_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the update user attribute request fails.
 */
export const UPDATE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the update user attribute request starts.
 */
export const UPDATE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the update user attribute request succeeds.
 */
export const UPDATE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the remove user attribute request fails.
 */
export const REMOVE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the remove user attribute request starts.
 */
export const REMOVE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the remove user attribute request succeeds.
 */
export const REMOVE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_SUCCESS';
