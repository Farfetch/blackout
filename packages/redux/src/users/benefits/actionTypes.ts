/**
 * Action type dispatched when the fetch benefits request fails.
 */
export const FETCH_USER_BENEFITS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_BENEFITS_FAILURE';
/**
 * Action type dispatched when the fetch benefits request starts.
 */
export const FETCH_USER_BENEFITS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_BENEFITS_REQUEST';
/**
 * Action type dispatched when the fetch benefits request succeeds.
 */
export const FETCH_USER_BENEFITS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_BENEFITS_SUCCESS';
