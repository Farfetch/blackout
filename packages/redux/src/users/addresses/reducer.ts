import * as actionTypes from './actionTypes';
import { AnyAction, combineReducers, Reducer } from 'redux';
import omit from 'lodash/omit';
import produce from 'immer';
import reducerFactory, {
  createReducerWithResult,
} from '../../helpers/reducerFactory';
import type {
  AddressEntity,
  AddressesEntity,
} from '../../entities/types/addresses.types';
import type {
  FetchUserAddressesSuccessAction,
  FetchUserAddressSuccessAction,
  UserAddressesState,
} from './types';
import type { StoreState } from '../../types';

export const INITIAL_STATE: UserAddressesState = {
  error: null,
  isLoading: false,
  result: null,
  addresses: {
    error: null,
    isLoading: false,
  },
  address: {
    error: {},
    isLoading: {},
  },
  /* Used for operations related with the default address that
    have a result associated, such as getting the default contact address */
  defaultAddressDetails: {
    error: null,
    isLoading: false,
    result: null,
  },
};

// Goes through the addresses list and returns the default `shipping|billing|contact` address details. The
// address returned is the one passed in `prop` param, i.e. isCurrentShipping, isCurrentBilling or isCurrentPreferred.
// If no default address with that prop exists, null is returned.
export const getDefaultAddress = (
  addressesList: AddressesEntity,
  prop: 'isCurrentShipping' | 'isCurrentBilling' | 'isCurrentPreferred',
): AddressEntity | null | undefined => {
  for (const key in addressesList) {
    const address = addressesList[key];

    if (address && address[prop]) {
      return address;
    }
  }
  return null;
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): UserAddressesState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS: {
      const fetchedAddressId = (action as FetchUserAddressSuccessAction).payload
        .result;

      if (!state) {
        return [fetchedAddressId];
      }

      if (!state.includes(fetchedAddressId)) {
        return [...state, fetchedAddressId];
      }

      return state;
    }
    case actionTypes.FETCH_USER_ADDRESSES_SUCCESS:
      return (action as FetchUserAddressesSuccessAction).payload.result;
    case actionTypes.CREATE_USER_ADDRESS_SUCCESS:
      if (!state) {
        return [action.meta.addressId];
      }

      return [...state, action.meta.addressId];
    case actionTypes.REMOVE_USER_ADDRESS_SUCCESS: {
      return (
        state && state.filter(addressId => addressId !== action.meta.addressId)
      );
    }
    default:
      return state;
  }
};

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): UserAddressesState['error'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_ADDRESSES_FAILURE:
    case actionTypes.CREATE_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_ADDRESSES_REQUEST:
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): UserAddressesState['isLoading'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ADDRESSES_REQUEST:
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return true;
    case actionTypes.FETCH_USER_ADDRESSES_FAILURE:
    case actionTypes.FETCH_USER_ADDRESSES_SUCCESS:
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS:
    case actionTypes.CREATE_USER_ADDRESS_FAILURE:
    case actionTypes.CREATE_USER_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.CREATE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const id = action.payload.result;
    const createdAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState) {
        draftState = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...createdAddress },
      };
    });
  },
  [actionTypes.UPDATE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const id = action.payload.result;
    const updatedAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState) {
        draftState = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...updatedAddress },
      };
    });
  },
  [actionTypes.REMOVE_USER_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    const currentAddresses = state?.addresses;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      draftState.addresses = omit(currentAddresses, addressId);

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    // Get prev default address so it can later be unmarked as the default
    const prevCurrentShippingAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentShipping',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentShippingAddress) {
        // Unmark previous shipping address as default
        const prevCurrentShippingAddressStore =
          addresses[prevCurrentShippingAddress.id];
        if (prevCurrentShippingAddressStore) {
          prevCurrentShippingAddressStore.isCurrentShipping = false;
        }
      }

      // Select the selected address as default
      const newDefaultShippingAddress = addresses[addressId];
      if (newDefaultShippingAddress) {
        newDefaultShippingAddress.isCurrentShipping = true;
      }

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;
    // Get prev default address so it can later be unmarked as the default
    const prevCurrentBillingAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentBilling',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentBillingAddress) {
        // Unmark previous billing address as default
        const prevCurrentBillingAddressStore =
          addresses[prevCurrentBillingAddress.id];
        if (prevCurrentBillingAddressStore) {
          prevCurrentBillingAddressStore.isCurrentBilling = false;
        }
      }

      // Select the selected address as default
      const newDefaultBillingAddress = addresses[addressId];
      if (newDefaultBillingAddress) {
        newDefaultBillingAddress.isCurrentBilling = true;
      }

      return draftState;
    });
  },
  [actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevCurrentContactAddress = getDefaultAddress(
      state?.addresses,
      'isCurrentPreferred',
    );

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      if (prevCurrentContactAddress) {
        // Unmark previous contact address as default
        const prevCurrentContactAddressStore =
          addresses[prevCurrentContactAddress.id];

        if (prevCurrentContactAddressStore) {
          prevCurrentContactAddressStore.isCurrentPreferred = false;
        }
      }

      // Select the selected address as default
      const newDefaultContactAddress = addresses[addressId];

      if (newDefaultContactAddress) {
        newDefaultContactAddress.isCurrentPreferred = true;
      }

      return draftState;
    });
  },
  [actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (
    state: NonNullable<StoreState['entities']>,
    action: AnyAction,
  ) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      if (!draftState) {
        return draftState;
      }

      const addresses = draftState.addresses;

      if (!addresses) {
        return draftState;
      }

      // Unmark the selected address as default
      const defaultContactAddress = addresses[addressId];

      if (defaultContactAddress) {
        defaultContactAddress.isCurrentPreferred = false;
      }

      return draftState;
    });
  },
};

export const address = (
  state = INITIAL_STATE.address,
  action: AnyAction,
): UserAddressesState['address'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ADDRESS_REQUEST:
      return {
        isLoading: { ...state.isLoading },
        error: {},
      };
    case actionTypes.FETCH_USER_ADDRESS_REQUEST:
    case actionTypes.UPDATE_USER_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: true,
        },
        error: {},
      };
    case actionTypes.FETCH_USER_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_USER_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
      };
    case actionTypes.FETCH_USER_ADDRESS_FAILURE:
    case actionTypes.UPDATE_USER_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
        error: {
          ...state.error,
          [action.meta.addressId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const addresses = reducerFactory(
  'FETCH_USER_ADDRESSES',
  INITIAL_STATE.addresses,
  actionTypes,
);

export const defaultAddressDetails = createReducerWithResult(
  'FETCH_USER_DEFAULT_CONTACT_ADDRESS',
  INITIAL_STATE.defaultAddressDetails,
  actionTypes,
);

export const getError = (
  state: UserAddressesState,
): UserAddressesState['error'] => state.error;
export const getIsLoading = (
  state: UserAddressesState,
): UserAddressesState['isLoading'] => state.isLoading;
export const getResult = (
  state: UserAddressesState,
): UserAddressesState['result'] => state.result;
export const getAddresses = (
  state: UserAddressesState,
): UserAddressesState['addresses'] => state.addresses;
export const getAddress = (
  state: UserAddressesState,
): UserAddressesState['address'] => state.address;
export const getDefaultAddressDetails = (
  state: UserAddressesState,
): UserAddressesState['defaultAddressDetails'] => state.defaultAddressDetails;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  addresses,
  address,
  defaultAddressDetails,
});

/**
 * Reducer for addresses state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const addressesReducer: Reducer<UserAddressesState> = (state, action) => {
  if (action.type === actionTypes.RESET_USER_ADDRESSES) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};

export default addressesReducer;
