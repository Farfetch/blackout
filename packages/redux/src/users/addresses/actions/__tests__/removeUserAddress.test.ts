import * as actionTypes from '../../actionTypes';
import {
  addressId2,
  mockUpdateAddressResponse,
  userId,
} from 'tests/__fixtures__/users';
import { deleteUserAddress } from '@farfetch/blackout-client';
import { INITIAL_STATE } from '../../reducer';
import { mockStore } from '../../../../../tests';
import { removeUserAddress } from '..';
import find from 'lodash/find';

jest.mock('@farfetch/blackout-client', () => ({
  ...jest.requireActual('@farfetch/blackout-client'),
  deleteUserAddress: jest.fn(),
}));

const addressesMockStore = (state = {}) =>
  mockStore({ addresses: INITIAL_STATE }, state);

const expectedConfig = undefined;
let store: ReturnType<typeof mockStore>;

describe('removeUserAddress() action creator', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    store = addressesMockStore({ entities: { user: { id: userId } } });
  });

  it('should create the correct actions for when the delete user address procedure fails', async () => {
    const expectedError = new Error('delete user address error');

    (deleteUserAddress as jest.Mock).mockRejectedValueOnce(expectedError);

    await expect(
      async () => await removeUserAddress(userId, addressId2)(store.dispatch),
    ).rejects.toThrow(expectedError);

    expect(deleteUserAddress).toHaveBeenCalledTimes(1);
    expect(deleteUserAddress).toHaveBeenCalledWith(
      { userId, id: addressId2 },
      expectedConfig,
    );
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        {
          meta: { addressId: addressId2 },
          type: actionTypes.REMOVE_USER_ADDRESS_REQUEST,
        },
        {
          type: actionTypes.REMOVE_USER_ADDRESS_FAILURE,
          payload: { error: expectedError },
          meta: { addressId: addressId2 },
        },
      ]),
    );
  });

  it('should create the correct actions for when the delete user address procedure is successful', async () => {
    (deleteUserAddress as jest.Mock).mockResolvedValueOnce(
      mockUpdateAddressResponse,
    );
    await removeUserAddress(userId, addressId2)(store.dispatch);

    const actionResults = store.getActions();

    expect(deleteUserAddress).toHaveBeenCalledTimes(1);
    expect(deleteUserAddress).toHaveBeenCalledWith(
      { userId, id: addressId2 },
      expectedConfig,
    );
    expect(actionResults).toMatchObject([
      {
        type: actionTypes.REMOVE_USER_ADDRESS_REQUEST,
      },
      {
        type: actionTypes.REMOVE_USER_ADDRESS_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.REMOVE_USER_ADDRESS_SUCCESS,
      }),
    ).toMatchSnapshot('delete user address success payload');
  });
});
