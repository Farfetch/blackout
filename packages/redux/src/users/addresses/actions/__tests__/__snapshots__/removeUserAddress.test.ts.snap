// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeUserAddress() action creator should create the correct actions for when the delete user address procedure is successful: delete user address success payload 1`] = `
Object {
  "meta": Object {
    "addressId": "2222222",
  },
  "type": "@farfetch/blackout-redux/REMOVE_USER_ADDRESS_SUCCESS",
}
`;
